lib/Jsrt/JsrtDebugManager.cpp

```diff
@@ -0,0 +1,693 @@
+                        index++;
+                    }
+                }
+            }
+        });
+    }
+
+    return scriptsArray;
+}
+
+Js::DynamicObject* JsrtDebugManager::GetSource(Js::ScriptContext* scriptContext, uint scriptId)
+{
+    Js::Utf8SourceInfo* utf8SourceInfo = nullptr;
+
+    for (Js::ScriptContext *tempScriptContext = this->threadContext->GetScriptContextList();
+    tempScriptContext != nullptr && utf8SourceInfo == nullptr && !tempScriptContext->IsClosed();
+        tempScriptContext = tempScriptContext->next)
+    {
+        tempScriptContext->GetSourceList()->MapUntil([&](int i, RecyclerWeakReference<Js::Utf8SourceInfo>* sourceInfoWeakRef) -> bool
+        {
+            Js::Utf8SourceInfo* sourceInfo = sourceInfoWeakRef->Get();
+            if (sourceInfo != nullptr && sourceInfo->IsInDebugMode() && sourceInfo->GetSourceInfoId() == scriptId)
+            {
+                utf8SourceInfo = sourceInfo;
+                return true;
+            }
+            return false;
+        });
+    }
+
+    Js::DynamicObject* sourceObject = nullptr;
+
+    if (utf8SourceInfo != nullptr)
+    {
+        sourceObject = (Js::DynamicObject*)Js::CrossSite::MarshalVar(utf8SourceInfo->GetScriptContext(), scriptContext->GetLibrary()->CreateObject());
+
+        JsrtDebugUtils::AddScriptIdToObject(sourceObject, utf8SourceInfo);
+        JsrtDebugUtils::AddFileNameOrScriptTypeToObject(sourceObject, utf8SourceInfo);
+        JsrtDebugUtils::AddLineCountToObject(sourceObject, utf8SourceInfo);
+        JsrtDebugUtils::AddPropertyToObject(sourceObject, JsrtDebugPropertyId::sourceLength, utf8SourceInfo->GetCchLength(), utf8SourceInfo->GetScriptContext());
+        JsrtDebugUtils::AddSouceToObject(sourceObject, utf8SourceInfo);
+    }
+
+    return sourceObject;
+}
+
+Js::JavascriptArray* JsrtDebugManager::GetStackFrames(Js::ScriptContext* scriptContext)
+{
+    if (this->stackFrames == nullptr)
+    {
+        this->stackFrames = Anew(this->GetDebugObjectArena(), JsrtDebugStackFrames, this);
+    }
+
+    return this->stackFrames->StackFrames(scriptContext);
+}
+
+bool JsrtDebugManager::TryGetFrameObjectFromFrameIndex(Js::ScriptContext *scriptContext, uint frameIndex, JsrtDebuggerStackFrame ** debuggerStackFrame)
+{
+    if (this->stackFrames == nullptr)
+    {
+        this->GetStackFrames(scriptContext);
+    }
+
+    return this->stackFrames->TryGetFrameObjectFromFrameIndex(frameIndex, debuggerStackFrame);
+}
+
+Js::DynamicObject* JsrtDebugManager::SetBreakPoint(Js::ScriptContext* scriptContext, Js::Utf8SourceInfo* utf8SourceInfo, UINT lineNumber, UINT columnNumber)
+{
+    Js::DebugDocument* debugDocument = utf8SourceInfo->GetDebugDocument();
+    if (debugDocument != nullptr && SUCCEEDED(utf8SourceInfo->EnsureLineOffsetCacheNoThrow()) && lineNumber < utf8SourceInfo->GetLineCount())
+    {
+        charcount_t charPosition;
+        charcount_t byteOffset;
```

Comment:
fixed

---
In reply to: [62756336](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62756336) [](ancestors = 62756336)